kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/partition/DynamicPartitioner.java

```diff
@@ -0,0 +1,256 @@
+      // decrease batch size back to the size that we actually checked.
+      decreaseBatchSize();
+      logger.debug("Best batch size found (all elements): {}", currentBatchSize);
+    }
+  }
+
+  /**
+   * writeAll intended for a generally stable currentBatchSize.
+   *
+   * <p>Write all the given elements to BigQuery in the smallest possible number of batches, as
+   * evenly as possible.
+   *
+   * <p>If we encounter a partitioning-related error, we will immediately lower the batch size and
+   * try again.
+   *
+   * <p>Every {@link #CONT_SUCCESS_COUNT_BUMP} calls, if there have been no errors, we will bump up
+   * the batch size.
+   */
+  private void establishedWriteAll(TableId table,
+                                   List<InsertAllRequest.RowToInsert> elements,
+                                   String topic,
+                                   Set<Schema> schemas) throws BigQueryConnectException,
+                                                               InterruptedException {
+    int currentIndex = 0;
+    while (currentIndex < elements.size()) {
+      try {
+        // very similar to equalPartitioner
+        // handle case where no partitioning is necessary:
+        if (elements.size() <= currentBatchSize) {
+          writer.writeRows(table, elements, topic, schemas);
+          currentIndex = elements.size();
+          return;
+        }
+
+        int numPartitions = (int)Math.ceil(elements.size() / (currentBatchSize * 1.0));
+        int minBatchSize = elements.size() / numPartitions;
+        int spareRows = elements.size() % numPartitions;
+
+        for (int partition = 0; partition < numPartitions; partition++) {
+          // the first spareRows partitions have an extra row in them.
+          int batchSize = partition < spareRows ? minBatchSize + 1 : minBatchSize;
+          int endIndex = currentIndex + batchSize;
+          writer.writeRows(table, elements.subList(currentIndex, endIndex), topic, schemas);
+          currentIndex = endIndex;
+        }
+      } catch (BigQueryException exception) {
+        if (isPartitioningError(exception)) {
+          // immediately decrease batch size and try again with remaining elements.
+          decreaseBatchSize();
+          contSuccessCount = 0;
+          logger.debug("Partition error during establishedWriteAll, reducing batch size to {}",
+                       currentBatchSize);
+          continue;
+        } else {
+          throw new BigQueryConnectException(
+              String.format("Failed to write to BigQuery table %s", table),
+              exception);
+        }
+      }
+    }
+
+    contSuccessCount++;
+    if (contSuccessCount >= CONT_SUCCESS_COUNT_BUMP) {
+      logger.debug("{} successful establishedWriteAlls in a row, increasing batchSize to {}",
+                   contSuccessCount, currentBatchSize);
+      increaseBatchSize();
+    }
+  }
+
+  private boolean isPartitioningError(BigQueryException exception) {
```

Comment:
Ah.. What I was thinking of was this:

> HTTP request size limit: 10 MB. Exceeding this value will cause invalid errors.

And:

> invalid	400	This error returns when there is any kind of invalid input other than an invalid query, such as missing required fields or an invalid table schema. Invalid queries return an invalidQuery error instead.	

So not 403, "real" 400 errors that don't come from their front end LBs. They don't send these now because these requests don't make it to the BQ part of the code, but if they "fix" the front ends (increase the max size), then we'd start getting these, I think. So, I think we should try and catch them in the same spot as the "empty" 400 responses.